File: lab/spec2.c
#define _GNU_SOURCE
#include <objc/runtime.h>
#include <objc/message.h>
#include <stdio.h>
#include <stdlib.h>



struct MyRawObject {
    void *isa;
    int x;
    int y;
};

void myMethod(struct MyRawObject * self, SEL _cmd) {
    printf("myMethod called! x = %d, y = %d\n", self->x, self->y);
}

int main() {
    const char *className = "MyDirtyClass";

    // Allocate class with space for 2 ints after isa (but no ivars formally declared)
    Class cls = objc_allocateClassPair(NULL, className, sizeof(struct MyRawObject) - sizeof(void *));
    objc_registerClassPair(cls);

    // Add method (no type encoding safety)
    class_addMethod(cls, sel_registerName("doSomething"), (IMP)myMethod, "");//, "v@:");

    // Create instance
    id obj = class_createInstance(cls, 0);

    // Cast to our dirty struct
    struct MyRawObject *hack = (struct MyRawObject *)obj;
    hack->x = 1337;
    hack->y = 4242;

    // Call method (normally!)
    objc_msgSend(obj, sel_getUid("doSomething"));

    // Cleanup
    object_dispose(obj);  // Normally handled by ARC or NSObject superclass
    return 0;
}

File: lab/spec2.m
#import <objc/runtime.h>
#import <objc/objc.h>
#import <objc/message.h>
#include <stdio.h>
#include <stdlib.h>

@interface Root
+ (instancetype) new;
- (void) delete;
@end

@implementation Root 
+ (instancetype) new {
    return class_createInstance(self, 0);

}
- (void) delete {
    object_dispose((id)self);
}
@end

@interface Name : Root {
    int ivar;
}
+ (instancetype) new;
- (Name*)initWithValue:(int)value and:(int)B;
- (void)printIvar;
- (void) delete;
@end

@implementation Name 

+ (instancetype) new {
    return [super new];
}

- (void) delete {
    
}

- (Name*)initWithValue:(int)value and:(int)B {
    self->ivar = value;
    return self;
}

- (void)printIvar {
    printf("Current ivar value = %d\n", self->ivar);
}

@end

int main(int argc, const char * argv[]) {
    //[Name helloWith:8];
    Name* defaultObj = [Name new];// initWithValue:123 and:8+8];
    [defaultObj printIvar]; 
    [defaultObj delete];
    return 0;
}
